packages/core/src/rpc/index.ts

```diff
@@ -432,23 +453,96 @@ export const createRpc = ({
   });
 
   let interval: NodeJS.Timeout | undefined;
+  let retryCount = 0;
 
   const rpc: Rpc = {
     // @ts-ignore
     request: queue.add,
-    subscribe({ onBlock, onError }) {
-      interval = setInterval(() => {
-        _eth_getBlockByNumber(rpc, { blockTag: "latest" })
-          .then(onBlock)
-          .catch(onError);
-      }, chain.pollingInterval);
-
-      common.shutdown.add(() => {
-        clearInterval(interval);
-      });
+    subscribe({ onBlock, onError: _onError }) {
+      if (wsTransport === undefined) {
+        interval = setInterval(() => {
+          _eth_getBlockByNumber(rpc, { blockTag: "latest" })
+            .then(onBlock)
+            .catch(_onError);
+        }, chain.pollingInterval);
+        common.shutdown.add(() => {
+          clearInterval(interval);
+        });
+      } else {
+        wsTransport
+          .value!.subscribe({
+            params: ["newHeads"],
+            onData: async (data) => {
+              if (data.result !== undefined) {
+                onBlock(data.result);
+                retryCount = 0;
+              } else if (data.error !== undefined) {
+                const error = data.error as Error;
+                console.log(`Websocket returned an error: ${error}`);
+
+                if (retryCount++ === RETRY_COUNT) {
+                  common.logger.warn({
+                    service: "rpc",
+                    msg: `Failed "eth_subscribe" subscription after ${retryCount + 1} consecutive errors. Switching to polling.`,
+                    error,
+                  });
+                  wsTransport = undefined;
```

Comment:
Should it unsubscribe before switching to polling? This also happens on line 505.